mqtt/mqtt3/src/client/connect.rs

```diff
@@ -16,7 +16,7 @@ where
     IoS: super::IoSource,
 {
     BeginBackOff,
-    EndBackOff(tokio::time::Delay),
+    EndBackOff(std::pin::Pin<Box<tokio::time::Sleep>>),
```

Comment:
It would be nice to avoid the allocation by changing the client code to support `State: !Unpin` (which would probably include making `Client: !Unpin`). And for `State` itself, the `pin-project` crate makes it easier to access `!Unpin` data of `!Unpin` enums.

Reply:
Yeah, I did one pass on using pin-project but got lost. I'm planning to do another pass again soon.

Reply:
Yes, it gets hairy when you have to re-assign `self` to be a different variant, and especially hairy when you have to re-assign `self` to be a different variant but containing some of the data from the current variant.